Update go-discover to support ECS discovery (#13782)
hashicorp/consul: website/content/docs/install/cloud-auto-join.mdx

- `addr_type` (optional) - the type of address to discover: `private_v4`, `public_v4`, `public_v6`. Default is `private_v4`. (>= 1.0)
- `access_key_id` (optional) - the AWS access key for authentication (see below for more information about authenticating).
- `secret_access_key` (optional) - the AWS secret access key for authentication (see below for more information about authenticating).
- `service` (optional) - The AWS service to filter. "ec2" or "ecs". Defaults to "ec2".
- `ecs_cluster` (optional) - The AWS ECS Cluster Name or Full ARN to limit searching within. Default none, search all.

> - `ecs_family` (optional) - The AWS ECS Task Definition Family to limit searching within. Default none, search all.

Reviewer: ```suggestion
- `ecs_family` (optional) - String value limits searches to a AWS ECS task definition family. By default, Consul searches all task definition families with the specified tags. 
```

> - `endpoint` (optional) - The endpoint URL of the AWS Service to use. If not set the AWS client will set this value, which defaults to the public DNS name for the service in the specified region.

Reviewer: ```suggestion
- `endpoint` (optional) - String value that specifies the endpoint URL of the AWS service to use. If not set, the AWS client sets the value, which defaults to the public DNS name for the service in the specified region.
```

- ECS task role metadata (container-specific).
- EC2 instance role metadata.

> To discover EC2 consul-servers, the only required IAM permission is `ec2:DescribeInstances`, and it is

Reviewer: ```suggestion
The only IAM permission required for discovering EC2 consul-servers is `ec2:DescribeInstances`. It is
```
Ideally, we'd start the second sentence with "We recommend . . . " so that it is in active voice per our style guide, but I don't think I can make those suggestions here.

recommended that you make a dedicated key used only to auto-join the datacenter. If the
region is omitted it will be discovered through the local instance's [EC2
metadata
endpoint](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html).

> To discover ECS consul-servers, the following IAM permissions are required on the AWS ECS Task Role associated with the Service performing discovery.
> - `ecs:ListClusters` (only used when `ecs_cluster` is not provided)
> - `ecs:ListServices` (only used when `ecs_cluster` is not provided)
> - `ecs:DescribeServices` (only used when `ecs_cluster` is not provided)
> - `ecs:ListTasks`
> - `ecs:DescribeTasks`
> If the region is omitted it will be discovered through the local instance's [ECS V4 metadata endpoint](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v4.html).
> 

Reviewer: ```suggestion
The AWS ECS task role associated with the service attempting to discover the `consul-server` must have the following IAM permissions:
- `ecs:ListClusters` (only used when `ecs_cluster` is not provided)
- `ecs:ListServices` (only used when `ecs_cluster` is not provided)
- `ecs:DescribeServices` (only used when `ecs_cluster` is not provided)
- `ecs:ListTasks`
- `ecs:DescribeTasks`
If the region is omitted from the configuration, Consul obtains it from the local instance's [ECS V4 metadata endpoint](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v4.html).

```
Is that introduction any clearer? Ignore if it doesn't seem like an improvement.
